bika/lims/catalog/catalog_multiplex_processor.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+
+from Acquisition import aq_base
+from bika.lims import api
+from bika.lims import logger
+from bika.lims.config import USE_COLLECTIVE_INDEXING
+from bika.lims.interfaces import IMultiCatalogBehavior
+from zope.interface import implements
+
+if USE_COLLECTIVE_INDEXING:
+    from collective.indexing.interfaces import IIndexQueueProcessor
+
+REQUIRED_CATALOGS = [
+    "auditlog_catalog",
+]
+
+
+class CatalogMultiplexProcessor(object):
+    """A catalog multiplex processor
+    """
+    if USE_COLLECTIVE_INDEXING:
+        implements(IIndexQueueProcessor)
+
+    def get_catalogs_for(self, obj):
+        catalogs = getattr(obj, "_catalogs", [])
+        for rc in REQUIRED_CATALOGS:
+            if rc in catalogs:
+                continue
+            catalogs.append(rc)
```

Comment:
Wouldn't the following be easier?:

```python
catalogs = getattr(obj, "_catalogs", []) + REQUIRED_CATALOGS
catalogs = list(set(catalogs))
return map(api.get_tool, catalogs)
```

Reply:
I think the current approach is more readable and performant, because in your approach the list hast to be iterated multiple times to generate first a `set` and convert it then to a `list`.
But indeed, it would be interesting what approach is faster...

... 5 minutes later ...

Here comes the proof:

```python
REQUIRED_CATALOGS = [
    "auditlog_catalog",
]

def v1(obj):
    catalogs = getattr(obj, "_catalogs", [])
    for rc in REQUIRED_CATALOGS:
        if rc in catalogs:
            continue
        catalogs.append(rc)
    return catalogs

def v2(obj):
     catalogs = getattr(obj, "_catalogs", []) + REQUIRED_CATALOGS
     catalogs = list(set(catalogs))
     return catalogs

class Content(object):
     _catalogs = ["setup_catalog"]

obj = Content()

if __name__ == '__main__':
    from timeit import timeit
    print timeit("v1(obj)", setup="from __main__ import obj, v1")
    print timeit("v2(obj)", setup="from __main__ import obj, v2")
```

Output:

```
0.572149038315
0.9786028862
```

Reply:
pheew!